internal/broker/broker.go

```diff
@@ -211,102 +209,140 @@ func (b *Broker) connectToOIDCServer(ctx context.Context) (*oidc.Provider, error
+	modesSupportedByUI := b.authModesSupportedByUI(supportedUILayouts)
+
+	for _, mode := range availableModes {
+		if !slices.Contains(modesSupportedByUI, mode) {
+			log.Infof(context.Background(), "Authentication mode %q is not supported by the UI", mode)
+			continue
+		}
+
+		authModesWithLabels = append(authModesWithLabels, map[string]string{
+			"id":    mode,
+			"label": authmodes.Label[mode],
 		})
 	}
 
-	if len(authModes) == 0 {
+	if len(authModesWithLabels) == 0 {
 		return nil, fmt.Errorf("no authentication modes available for user %q", session.username)
 	}
 
-	session.authModes = availableModes
-	if err := b.updateSession(sessionID, session); err != nil {
-		return nil, err
-	}
-
-	return authModes, nil
+	return authModesWithLabels, nil
 }
 
-func (b *Broker) supportedAuthModesFromLayout(supportedUILayouts []map[string]string) (supportedModes map[string]string) {
-	supportedModes = make(map[string]string)
-	for _, layout := range supportedUILayouts {
-		supportedEntries := strings.Split(strings.TrimPrefix(layout["entry"], "optional:"), ",")
-		switch layout["type"] {
-		case "qrcode":
-			if !strings.Contains(layout["wait"], "true") {
+func (b *Broker) availableAuthModes(session session) (availableModes []string, err error) {
+	if len(session.nextAuthModes) > 0 {
+		for _, mode := range session.nextAuthModes {
+			if !authModeIsAvailable(session, mode) {
+				log.Infof(context.Background(), "Authentication mode %q is not available", mode)
 				continue
 			}
-			deviceAuthID := authmodes.DeviceQr
-			if layout["renders_qrcode"] == "false" {
-				deviceAuthID = authmodes.Device
-			}
-			supportedModes[deviceAuthID] = "Device Authentication"
+			availableModes = append(availableModes, mode)
+		}
+		return availableModes, nil
+	}
 
-		case "form":
-			if slices.Contains(supportedEntries, "chars_password") {
-				supportedModes[authmodes.Password] = "Local Password Authentication"
-			}
+	switch session.mode {
+	case sessionmode.ChangePassword, sessionmode.ChangePasswordOld:
+		// session is for changing the password
+		if !tokenExists(session) {
+			return nil, errors.New("user has no cached token")
+		}
+		return []string{authmodes.Password}, nil
 
-		case "newpassword":
-			if slices.Contains(supportedEntries, "chars_password") {
-				supportedModes[authmodes.NewPassword] = "Define your local password"
+	default:
+		// session is for login
+		modes := append(b.provider.SupportedOIDCAuthModes(), authmodes.Password)
+		for _, mode := range modes {
+			if authModeIsAvailable(session, mode) {
+				availableModes = append(availableModes, mode)
 			}
 		}
+		return availableModes, nil
 	}
+}
 
+func authModeIsAvailable(session session, authMode string) bool {
+	switch authMode {
+	case authmodes.Password:
+		return tokenExists(session)
+	case authmodes.NewPassword:
+		return true
+	case authmodes.Device, authmodes.DeviceQr:
+		return session.oidcServer != nil && session.oidcServer.Endpoint().DeviceAuthURL != "" && !session.isOffline
```

Comment:
@denisonbarbosa Regarding https://github.com/ubuntu/authd-oidc-brokers/pull/349#discussion_r1959880851, do you think we should extract this line into a method of the `Provider` interface?

Reply:
Great, I fully agree. Anything else missing on this PR for your approval?